openfisca_core/parameters.py

```diff
@@ -462,10 +462,15 @@ def __init__(self, name, node, instant_str):
         self._name = name
         self._instant_str = instant_str
         self._children = {}
+
+        # if trace:
+        self._original_children = {}
```

Comment:
Why are changes in `ParameterNodeAtInstant` required ? I thought a new class would be self-contained.

Reply:
Basically because prior to this change ParameterNodeAtInstant didn't keep track of the actual `ParameterNode`s, `Parameter`s or `Scale`s that are its children, but only of the values attached to these children at its instant (the method `get_at_instant` is improperly named, it should really be `get_value_at_instant`). So if we didn't make the above change we could not get access to a parameter's name in `record_calculation_parameter_access`, only to its value.

Also see [this comment](https://github.com/openfisca/openfisca-core/pull/756#issuecomment-438006702) for more discussion of some design issues in this area.

Reply:
:thinking: from my point of view (pretty far to be honest this links a production class to a debug one).

But if only the names is required, can't
`record_calculation_parameter_access(self, parameter, period):` be refactored in
`record_calculation_parameter_access(self, parameter_value, parameter_name, period):`
and `parameter_name` be provided from the `key` argument of `def __getattr__(self, key):` ?

Reply:
It's a possibility. The original intent was to get at the "effective period" of a parameter (as opposed to the requested period); that proved very difficult and we gave up on that.

Reply:
Strongly support @guillett's point. Adding `parameter_name` to the signature of `record_calculation_parameter_access` seems more elegant.

Reply:
I tried to pass the tuple:

```python
        for child in node.children.items():
            child_at_instant = child[1]._get_at_instant(instant_str)
            if child_at_instant is not None:
                self._children[child_name] = child + (child_at_instant,)
                # ...
```

But it wreaks havoc...